[intesis] New Binding for Intesis WiFi adapters used in various AC systems

hmerk opened the pull request:
This Version supports IntesisHome Devices only. It is planned to integrate IntesiBox support as well.

hmerk commented:
Hmm, could someone explain to me why Travis is failing ????

Hilbrand commented:
You miss the binding in the bundle/pom.xml

hmerk commented:
Dumb me, thanks @Hilbrand for pointing me to the obvious.

hmerk commented:
Ok, fine with Java 8, failing on Java 11

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,237 @@
+        deviceIp = config.ipAddress;
+        password = config.password;
+        try {
+            httpClient.start();
```
Better use `HttpClientFactory` and us `getCommonHttpClient` to get a httpclient. See other bindings on how it's used.

hmerk replied:
Thanks @Hilbrand for pointing me into this direction. Will refactor my binding to address your comment.

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                    break;
+                case TEMP_CHANNEL:
+                    uid = 9;
+                    value = (Integer.parseInt(command.toString().replace(" °C", ""))) * 10;
```
As you use Units of Measure, you can cast this to `QuantityType` and use `toUnit()` to convert it to celsius. Your code would fail if the user uses Fahrenheit.

hmerk replied:
No, the command has to deliver just an integer, so I have to remove the unit symvol and multiply the interger value by 10.

Hilbrand replied:
```suggestion
                    if (command instanceof QuantityType) {
                        value = ((QuantityType) command).toUnit(SIUnits.CELSIUS).intValue()  * 10;
                    }
```

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+            String contentString = "{\"command\":\"setdatapointvalue\",\"data\":{\"sessionID\":\"" + sessionId
+                    + "\", \"uid\":" + uid + ",\"value\":" + value + "}}";
+            String response = api.postRequest(ipAddress, contentString);
+            if (response != null && !response.isEmpty()) {
```
See above

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+    public void getAllUidValues() {
+        logger.debug("Polling IntesisHome device for actuall status");
+        if (thing.getStatus() != ThingStatus.ONLINE) {
+            String sessionString = "{\"command\":\"login\",\"data\":{\"username\":\"Admin\",\"password\":\"" + password
+                    + "\"}}";
+            String response = api.postRequest(ipAddress, sessionString);
+            if (response != null && !response.isEmpty()) {
+                boolean success = IntesisHomeJSonDTO.getSuccess(response);
+                if (success) {
```
It might be a good approach to extract this code into a method and re-use it.

hmerk replied:
Don't see the real benefit.

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+            } else {
+                JsonElement dpvalNode = IntesisHomeJSonDTO.getData(response).get("dpval");
+                dpval[] uid = gson.fromJson(dpvalNode, dpval[].class);
+                int value = uid[0].value;
```
Better check if the array has the length you expect.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                updateState(WINDSPEED_CHANNEL, stateValue);
+                stateValue = new DecimalType(uid[3].value);
+                updateState(SWINGUD_CHANNEL, stateValue);
+                int unit = (uid[4].value) / 10;
+                stateValue = QuantityType.valueOf(unit, SIUnits.CELSIUS);
```
If I see correctly, the positions after the decimal point will be truncated. If this is intended, better use `Math.round()`, as integer division will always rounding down.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/binding/binding.xml:
```diff
@@ -0,0 +1,10 @@
+	xsi:schemaLocation="https://openhab.org/schemas/binding/v1.0.0 https://openhab.org/schemas/binding-1.0.0.xsd">
+
+	<name>Intesis Binding</name>
+	<description>This is the binding for Intesis.</description>
```
Could you be a bit more verbose, so that somebody who hasn't heard about it gets a clue what this is about?

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/i18n/intesis.properties:
```diff
@@ -0,0 +1,77 @@
+# FIXME: please substitute the xx_XX with a proper locale, ie. de_DE
+# FIXME: please do not add the file to the repo if you add or change no content
```
Can this be removed?

hmerk replied:
removed

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,104 @@
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<!-- Sample Thing Type -->
```
Can this be removed?

hmerk replied:
Removed

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,104 @@
+		</channels>
+
+		<config-description>
+			<parameter name="password" type="text" required="true">
+				<label>Password</label>
+				<description>Password for API access.</description>
+				<context>password</context>
+			</parameter>
+			<parameter name="ipAddress" type="text" required="true">
+				<label>IP Address</label>
+				<description>IP Address of the IntesisHome device.</description>
+				<context>network-address</context>
+			</parameter>
```
I think it's more common if the IP address is entered first and after that, the password.

hmerk replied:
Changed

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,104 @@
+		<description>@text/channel-type.intesis.mode.description</description>
+		<state>
+			<options>
+				<option value="0">Auto</option>
+				<option value="1">Heat</option>
+				<option value="2">Dry</option>
+				<option value="3">Fan</option>
+				<option value="4">Cool</option>
```
If I see correctly, this doesn't match with the i18n file. As it is redundant, you could remove it in the i18n file.

hmerk replied:
Translations are work in progress

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,104 @@
+	</channel-type>
+	<channel-type id="outdoorTemp">
+		<item-type>Number:Temperature</item-type>
+		<label>@Outdoor Temperature</label>
```
```suggestion
		<label>Outdoor Temperature</label>
```

hmerk replied:
Done

fwolter commented on bundles/pom.xml:
```diff
@@ -1,4 +1,5 @@
-<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Can you leave the header as-is?

hmerk replied:
Don't know where this commes fromm, as my local file is ok.

fwolter commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,83 @@
+| Channel ID | Item Type          | Description                                                           | Possible Values |
+|------------|--------------------|-----------------------------------------------------------------------|-|
+| power      | Switch             | Turns power on/off for your climate system.                           | ON, OFF |
+| mode       | String             | The heating/cooling mode.                                             | AUTO, HEAT, COOL, DRY, FAN |
+| windspeed  | String             | Fan speed (if applicable)                                             | AUTO, 1-4 |
```
Channels are handled as `Number` in your code. You might want to specify which number maps to which mode. Same for `swingUpDown`.

hmerk replied:
Changed to Number, possible values accordingly. There will be no mapping explanation, as different AC Units may have different order of modes.

fwolter commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,83 @@
+
+## Channels
+
+| Channel ID | Item Type          | Description                                                           | Possible Values |
```
```suggestion
| Channel ID | Channel Type          | Description                                                           | Possible Values |
```

hmerk replied:
No, this describes that the channel has to be linked to a certain item Type. This is common usage in the readme files.

fwolter reviewed (changes requested):
I reviewed your code, here is my feedback.

Great that you applied spotless before filing this PR!

There are some checkstyle warnings left. You could take a look at `target/code-analysis/report.html`.

hmerk commented:
Thanks for your first review @fwolter did not expect it at this stage.
I marked this PR as WIP as I know there is still some work to do.
But as said, thanks, your comments are welcome and I will integrate your change requests.

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+        int value = 0;
+        String channelId = channelUID.getId();
+        if (command instanceof RefreshType) {
+            // The thing is updated by the scheduled automatic refresh so do nothing here.
```
I could, but it is a really bad idea as I saw in the debugger, that all channels are refreshed at startup, which causes unnecessary traffic.

fwolter replied:
You could start a timer of e.g. 3 sec. for delaying the polling. Any Refresh command re-triggers it, so that only the last poll is invoked.

hmerk replied:
Don't see any benefit, as we poll the status every 30 seconds....

hmerk commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,83 @@
+| mode       | String             | The heating/cooling mode.                                             | AUTO, HEAT, COOL, DRY, FAN |
+| windspeed  | String             | Fan speed (if applicable)                                             | AUTO, 1-4 |
+| temperature | Number:Temperature | The currently set target temperature.                                 | |
+| returnTemp | Number:Temperature | (Readonly) The ambient air temperature.                               | |
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,83 @@
+
+## Full Example
+
+The binding can be fully setup from the Paper UI but if you decide to use files here is a full example:
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/IntesisHandlerFactory.java:
```diff
@@ -0,0 +1,75 @@
+
+    @Activate
+    public IntesisHandlerFactory(@Reference HttpClientFactory httpClientFactory, ComponentContext componentContext) {
+        logger.debug("Activate Shelly HandlerFactory");
```
Removed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisHomeHttpApi.java:
```diff
@@ -0,0 +1,73 @@
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link IntesisHomeHttpApi} wraps the IntesisHome REST API and provides various low level function to access the
+ * device api (not
+ * cloud api).
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+public class IntesisHomeHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(IntesisHomeHandler.class);
+    // private final HttpClient httpClient;
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+
+    private @Nullable ScheduledFuture<?> refreshJob;
+
+    private int refreshInterval = 30;
```
Changed and moved to binding constants

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+    private String password = "";
+    private String sessionId = "";
+
+    Gson gson = new Gson();
```
Made Gson final

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
```
Hmm, You are commenting on code created with the binding skeleton, why ?
It is also common usage to have this debug logger.

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
+        updateStatus(ThingStatus.ONLINE, ThingStatusDetail.CONFIGURATION_PENDING);
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                        contentString = "{\"command\":\"login\",\"data\":{\"username\":\"Admin\",\"password\":\""
+                                + password + "\"}}";
+                        response = api.postRequest(ipAddress, contentString);
+                        if (response != null && !response.isEmpty()) {
```
Removed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                }
+            } catch (Exception e) {
+            }
+        }, 2, TimeUnit.SECONDS);
```
No, changed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+            refreshJob.cancel(true);
+            this.refreshJob = null;
+        }
+        try {
+            String contentString = "{\"command\":\"logout\",\"data\":{\"sessionID\":\"" + sessionId + "\"}}";
+            api.postRequest(ipAddress, contentString);
+        } catch (Exception e) {
+        }
```
Removed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                    uid = 5;
+                    value = Integer.parseInt(command.toString());
+                    break;
+                case SWINGLR_CHANNEL:
```
Leftover, removed

hmerk commented:
@fwolter I applied most of your change requests, some comments left.
Localisation ist still work in Progress.
Thank's again for your valuable input.

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisHomeHttpApi.java:
```diff
@@ -0,0 +1,73 @@
+            if (response != null && !response.isEmpty()) {
+                return response;
+            }
+        } catch (Exception e) {
+        }
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,276 @@
+                        updateStatus(ThingStatus.OFFLINE);
+                    }
+                }
+            } catch (Exception e) {
```
Removed

hmerk commented:
@fwolter @Hilbrand I have finished my work in progress and will be ready for your reviews.
What has changed:
- Finished Localisation
- Added dynamic channels and stateOptions based on device response
- Changed channel stateOptions from numbers to strings (better fit with homeKit, Alexa or GoogleAssistant)
- changed session handling.

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+    @Override
+    public void initialize() {
+        logger.debug("Start initializing!");
+        updateStatus(ThingStatus.UNKNOWN, ThingStatusDetail.CONFIGURATION_PENDING);
```
```suggestion
        updateStatus(ThingStatus.UNKNOWN);
```
See https://www.openhab.org/docs/concepts/things.html#status-details

hmerk replied:
Removed

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/gson/IntesisHomeJSonDTO.java:
```diff
@@ -0,0 +1,185 @@
+ */
+public class IntesisHomeJSonDTO {
+
+    public static class response {
```
Classes should start with uppercase character:
```suggestion
    public static class Response {
```

hmerk replied:
Changed

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+                        updateStatus(ThingStatus.OFFLINE);
+                    }
+                }
+            } finally {
```
Remove empty finally

hmerk replied:
Done

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+        }
+
+        // start background initialization:
+        scheduler.schedule(() -> {
```
This is a horribly long inline method. Please put this code in a method and pass with `this::startupInitialization` (or whatever name). Also do split the content of this method in multiple smaller methods. You can select a block of code in eclipse (if you use that) and `Ctrl-1` will give you if you select a coherent part the option to extract as method. Even better would be to split different code behavior into separate classes.

hmerk replied:
Done

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+                    break;
+                case CHANNEL_TYPE_VANESUD:
+                    uid = 5;
+                    if (("AUTO").equals(command.toString())) {
+                        value = 0;
+                    } else if (("SWING").equals(command.toString())) {
+                        value = 10;
+                    } else if (("SWIRL").equals(command.toString())) {
+                        value = 11;
+                    } else if (("WIDE").equals(command.toString())) {
+                        value = 12;
+                    } else {
+                        value = Integer.parseInt(command.toString());
+                    }
```
This can be a switch statement

hmerk replied:
Done

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang.WordUtils;
```
Remove usage of apache library as it will be removed in openHAB 3

hmerk replied:
Removed

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,566 @@
+ * @author Hans-Jörg Merk - Initial contribution
+ */
+@NonNullByDefault
+public class IntesisHomeHandler extends BaseThingHandler implements DynamicStateDescriptionProvider {
```
Put `DynamicStateDescriptionProvider` in a separate class and let it extend `BaseDynamicStateDescriptionProvider` that way you can make a minimal implementation.

hmerk replied:
Removed Leftover

hmerk commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,87 @@
+
+## Thing Configuration
+
+The binding needs two configuration parameters, passwort and IP-Address.
```
Changed

hmerk commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,87 @@
+## Channels
+
+| Channel ID | Item Type          | Description                                                           | Possible Values |
+|------------|--------------------|-----------------------------------------------------------------------|-|
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisHomeHttpApi.java:
```diff
@@ -0,0 +1,76 @@
+    public String postRequest(String ipAddress, String contentString) {
+        String url = "http://" + ipAddress + "/api.cgi";
+        try {
+            Request request = httpClient.POST(url);
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/gson/IntesisHomeJSonDTO.java:
```diff
@@ -0,0 +1,185 @@
+public class IntesisHomeJSonDTO {
+
+    public static class response {
+        @SerializedName("success")
```
Removed all occurences

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/IntesisDynamicStateDescriptionProvider.java:
```diff
@@ -0,0 +1,63 @@
+ */
+@Component(service = { DynamicStateDescriptionProvider.class, IntesisDynamicStateDescriptionProvider.class })
+@NonNullByDefault
+public class IntesisDynamicStateDescriptionProvider implements DynamicStateDescriptionProvider {
```
Changed, but do not see what code can be removed .

Hilbrand replied:
You don't have to maintain the `descriptions`. The base class provides this functionallity. So all methods related to maintaining that field are not needed anymore.

hmerk replied:
Tried to remove 
removeAllDescriptions();
getStateDescription();
But this breaks functionality. PaperUI does not show the Options anymore

hmerk commented:
Thanks @Hilbrand for your review and your comments which have been addressed (nearly) completely.  Only one question left so far.

Hilbrand commented:
Best to rebase `git pull rebase orgin/2.5.x` (assuming origin points to this repo) and the `git push --force-with-lease` back to github to get it insync.

Hilbrand commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/IntesisDynamicStateDescriptionProvider.java:
```diff
@@ -0,0 +1,65 @@
+ */
+@Component(service = { DynamicStateDescriptionProvider.class, IntesisDynamicStateDescriptionProvider.class })
+@NonNullByDefault
+public class IntesisDynamicStateDescriptionProvider extends BaseDynamicStateDescriptionProvider {
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private final Map<ChannelUID, @Nullable StateDescription> descriptions = new ConcurrentHashMap<>();
+
+    public void setDescription(ChannelUID channelUID, @Nullable StateDescription description) {
+        logger.debug("Adding state description for channel {}", channelUID);
+        descriptions.put(channelUID, description);
+    }
+
+    public void removeAllDescriptions() {
+        logger.debug("Removing all state descriptions");
+        descriptions.clear();
+    }
+
+    @Override
+    public @Nullable StateDescription getStateDescription(Channel channel,
+            @Nullable StateDescription originalStateDescription, @Nullable Locale locale) {
+        StateDescription description = descriptions.get(channel.getUID());
+        return description;
+    }
+
+    @Override
+    @Deactivate
+    public void deactivate() {
+        descriptions.clear();
+    }
+}
```
```suggestion
public class IntesisDynamicStateDescriptionProvider extends BaseDynamicStateDescriptionProvider {
}
```

And in `IntesisHomeHandler` replace
```
    updateStateDescription(thing.getChannel(channelId).getUID(), stateOptions);
```
with
```
     intesisStateDescriptionProvider.setStateOptions(thing.getChannel(channelId).getUID(), stateOptions);
```

And remove `updateStateDescription`.

hmerk replied:
Thanks @Hilbrand, I will give it a try later tonight or tomorrow and report back.

hmerk replied:
Thanks again @Hilbrand !!!! That worked an is implemented now. Had some trouble with the commit, but solved it.

Hilbrand reviewed (approved):
LGTM

hmerk commented:
Thanks @Hilbrand !!!!!

@fwolter as you already started a review earlier, could you please do a second review, so we could probably get this merged for 2.5.9 ?

hmerk commented:
@fwolter ping ;-)

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,552 @@
+                case CHANNEL_TYPE_TARGETTEMP:
+                    uid = 9;
+                    if (command instanceof QuantityType) {
+                        value = ((QuantityType) command).toUnit(SIUnits.CELSIUS).intValue() * 10;
```
`toUnit()` is Nullable. Please add a null check.

hmerk replied:
Could you help me with this ??? Can't figure out how to detect if toUnit(Unit) can be checked here.

hmerk replied:
Solved and changed

fwolter commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,552 @@
+        }
+    }
+
+    @SuppressWarnings("null")
```
Please add a null check instead of suppressing the warning.

hmerk replied:
Done

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/i18n/intesis_de.properties:
```diff
@@ -0,0 +1,42 @@
+channel-type.intesis.mode.state.option.dry = Trocknen
+channel-type.intesis.mode.state.option.fan = Ventilation
+channel-type.intesis.mode.state.option.heat = Heizen
+channel-type.intesis.targetTemperature.label = Soll Temperatur
```
Same for below.
```suggestion
channel-type.intesis.targetTemperature.label = Solltemperatur
```

hmerk replied:
Corrected

fwolter commented on bundles/org.openhab.binding.intesis/src/main/resources/ESH-INF/i18n/intesis_de.properties:
```diff
@@ -0,0 +1,42 @@
+channel-type.intesis.targetTemperature.description = Stellt die gew�nschte Soll Temperatur ein.
+channel-type.intesis.ambientTemperature.label = Umgebungstemperatur
+channel-type.intesis.ambientTemperature.description = Zeigt die aktuelle Raumtemperatur an.
+channel-type.intesis.outdoorTemperature.label = Au�enemperatur
```
```suggestion
channel-type.intesis.outdoorTemperature.label = Au�entemperatur
```

hmerk replied:
Corrected

hmerk commented on bundles/org.openhab.binding.intesis/README.md:
```diff
@@ -0,0 +1,93 @@
+
+
+
+###Note that individual A/C units may not support all channels, or all possible values for those channels.
```
Changed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,552 @@
+        int value = 0;
+        String channelId = channelUID.getId();
+        if (command instanceof RefreshType) {
+            // getAllUidValues();
```
Removed debug logger.
Refresh command is not supported as the binding polls all values every 30 seconds.

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/api/IntesisHomeHttpApi.java:
```diff
@@ -0,0 +1,77 @@
+                return null;
+            }
+        } catch (TimeoutException | InterruptedException | ExecutionException e) {
+            logger.debug("Could not make HTTP Post request", e);
```
removed logging of stack trace

hmerk commented:
Hmmm, seems that Travis failure is not related to Intesis binding: 
SEVERE: Step 'eclipse wtp formatters - XML' found problem in 'bundles/org.openhab.binding.heos/src/main/feature/feature.xml':
java.lang.ClassNotFoundException: com.diffplug.spotless.extra.eclipse.wtp.EclipseXmlFormatterStepImpl

hmerk commented:
@fwolter @Hilbrand 
Travis is still failing on heos binding Java 8, Java 11 is fine and Jenkins failed, but Intesis binding was green.
So in general, there seems no issue with the Intesis binding.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,557 @@
+                    Data data = gson.fromJson(resp.data.toString(), Data.class);
+                    Dp dp = gson.fromJson(data.dp.toString(), Dp.class);
+                    Datapoints[] datapoints = gson.fromJson(dp.datapoints, Datapoints[].class);
+                    for (int i = 0; i < datapoints.length; i++) {
+                        Descr descr = gson.fromJson(datapoints[i].descr, Descr.class);
```
A for-each loop would be preferred here unless you really need to know the current index.
```suggestion
                    for (Datapoints datapoint : datapoints) {
                        Descr descr = gson.fromJson(datapoint.descr, Descr.class);
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,557 @@
+                        switch (datapoints[i].uid) {
+                            case 2:
+                                List<String> opModes = new ArrayList<>();
+                                String[] modString = descr.states;
+                                for (int i1 = 0; i1 < modString.length; i1++) {
+                                    switch (modString[i1]) {
```
```suggestion
                                for (String modString : descr.states) {
                                    switch (modString) {
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,557 @@
+                                break;
+                            case 4:
+                                List<String> fanLevels = new ArrayList<>();
+                                String[] fanString = descr.states;
+                                for (int i1 = 0; i1 < fanString.length; i1++) {
+                                    if ("AUTO".contentEquals(fanString[i1])) {
+                                        fanLevels.add("AUTO");
+                                    } else {
+                                        fanLevels.add(fanString[i1]);
+                                    }
+                                }
```
```suggestion
                                for (String fanString : descr.states) {
                                    if ("AUTO".contentEquals(fanString)) {
                                        fanLevels.add("AUTO");
                                    } else {
                                        fanLevels.add(fanString);
                                    }
                                }
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,557 @@
+                                break;
+                            case 5:
+                                List<String> swingUDModes = new ArrayList<>();
+                                String[] swingUDString = descr.states;
+                                for (int i1 = 0; i1 < swingUDString.length; i1++) {
+                                    if ("AUTO".contentEquals(swingUDString[i1])) {
+                                        swingUDModes.add("AUTO");
+                                    } else if ("10".contentEquals(swingUDString[i1])) {
+                                        swingUDModes.add("SWING");
+                                    } else if ("11".contentEquals(swingUDString[i1])) {
+                                        swingUDModes.add("SWIRL");
+                                    } else if ("12".contentEquals(swingUDString[i1])) {
+                                        swingUDModes.add("WIDE");
+                                    } else {
+                                        swingUDModes.add(swingUDString[i1]);
+                                    }
+                                }
```
```suggestion
                                for (String swingUDString : descr.states) {
                                    if ("AUTO".contentEquals(swingUDString)) {
                                        swingUDModes.add("AUTO");
                                    } else if ("10".contentEquals(swingUDString)) {
                                        swingUDModes.add("SWING");
                                    } else if ("11".contentEquals(swingUDString)) {
                                        swingUDModes.add("SWIRL");
                                    } else if ("12".contentEquals(swingUDString)) {
                                        swingUDModes.add("WIDE");
                                    } else {
                                        swingUDModes.add(swingUDString);
                                    }
                                }
```

hmerk replied:
Refactored

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,557 @@
+                                break;
+                            case 6:
+                                List<String> swingLRModes = new ArrayList<>();
+                                String[] swingLRString = descr.states;
+                                for (int i1 = 0; i1 < swingLRString.length; i1++) {
+                                    if ("AUTO".contentEquals(swingLRString[i1])) {
+                                        swingLRModes.add("AUTO");
+                                    } else if ("10".contentEquals(swingLRString[i1])) {
+                                        swingLRModes.add("SWING");
+                                    } else if ("11".contentEquals(swingLRString[i1])) {
+                                        swingLRModes.add("SWIRL");
+                                    } else if ("12".contentEquals(swingLRString[i1])) {
+                                        swingLRModes.add("WIDE");
+                                    } else {
+                                        swingLRModes.add(swingLRString[i1]);
+                                    }
+                                }
```
It looks like this code is almost identical to the code above. Can you refactor this code so it isn't duplicated?

hmerk replied:
Refactored

hmerk commented:
Thanks @cpmeister for your valuable comments. I have addressed all points.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/gson/IntesisHomeJSonDTO.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public static class Dp {
+        public JsonArray datapoints; // dataPoints
```
Any reason that you aren't using classes for the nested json structure?
```suggestion
        public Datapoints[] datapoints; // dataPoints
```

hmerk replied:
I tried using classes, but the returned json is a bit malformed, so this was the simplest way to solve it.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+        return response;
+    }
+
+    public boolean getSuccess(String response) {
+        Response resp = gson.fromJson(response, Response.class);
+        boolean success = resp.success;
+        return success;
+    }
+
+    public void getInfo() {
+        String contentString = "{\"command\":\"getinfo\",\"data\":\"\"}";
+        String response = api.postRequest(ipAddress, contentString);
+        logger.trace("getInfo response : {}", response);
+        if (response != null) {
+            boolean success = getSuccess(response);
+            if (success) {
+                Response resp = gson.fromJson(response, Response.class);
+                Data data = gson.fromJson(resp.data.toString(), Data.class);
```
It looks like you are parsing the response multiple times here, first to check the success flag, then again if the flag was true. Can you optimize this so it only parses the response a single time?

hmerk replied:
removed getSuccess and avoided double parsing

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+        return response;
+    }
+
+    public boolean getSuccess(String response) {
+        Response resp = gson.fromJson(response, Response.class);
+        boolean success = resp.success;
+        return success;
+    }
```
I don't think you should have this method, parsing is too expensive to throw away the result on success. It would be better if you inlined this code in the callers and then got rid of this method.

hmerk replied:
removed

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+                + "\"}}";
+        String response = api.postRequest(ipAddress, contentString);
+        if (response != null) {
+            Response resp = gson.fromJson(response, Response.class);
```
You should catch any `JsonSyntaxException` that might be thrown when using gson. Sometimes vendors accidentally break their own api. You should make sure that your code can appropriately handle when that happens.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+                    break;
+                case CHANNEL_TYPE_VANESLR:
+                    uid = 6;
+                    switch (command.toString()) {
+                        case "AUTO":
+                            value = 0;
+                            break;
+                        case "1":
+                        case "2":
+                        case "3":
+                        case "4":
+                        case "5":
+                        case "6":
+                        case "7":
+                        case "8":
+                        case "9":
+                            value = Integer.parseInt(command.toString());
+                            break;
+                        case "SWING":
+                            value = 10;
+                            break;
+                        case "SWIRL":
+                            value = 11;
+                            break;
+                        case "WIDE":
+                            value = 12;
+                            break;
+                    }
```
Please refactor this too.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+        return success;
+    }
+
+    public void getInfo() {
```
Starting a method name with `get` implies that the method would return something, but this code doesn't do that. I would suggest renaming this method `populateProperties`.

hmerk replied:
renamed

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+            thingBuilder.withChannel(channel);
+            updateThing(thingBuilder.build());
+
+            if (getThing().getChannel(channelId) != null) {
```
You add this channel to the thing in the code preceding this, so would this ever be false?

hmerk replied:
You're right, removed superfluend null check

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+                                    }
+                                }
+                                switch (datapoint.uid) {
+                                    case 5:
+                                        channelId = CHANNEL_TYPE_VANESUD;
+                                        properties.put("Supported vane up/down modes", swingModes.toString());
+                                        updateProperties(properties);
+                                        addChannel(channelId, itemType, swingModes);
+                                        break;
+                                    case 6:
+                                        channelId = CHANNEL_TYPE_VANESUD;
+                                        properties.put("Supported vane up/down modes", swingModes.toString());
+                                        updateProperties(properties);
+                                        addChannel(channelId, itemType, swingModes);
+                                        break;
```
Looks like a copy/paste bug here.

hmerk replied:
Indeed, changed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+                logout();
+                Response resp = gson.fromJson(response, Response.class);
+                Data data = gson.fromJson(resp.data.toString(), Data.class);
+                Dpval[] dpval = gson.fromJson(data.dpval, Dpval[].class);
+                for (int i = 0; i < dpval.length; i++) {
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+                            }
+                            break;
+                        case 5:
+                            if ((dpval[i].value) == 0) {
+                                updateState(CHANNEL_TYPE_VANESUD, StringType.valueOf("AUTO"));
+                            } else if ((dpval[i].value) == 10) {
+                                updateState(CHANNEL_TYPE_VANESUD, StringType.valueOf("SWING"));
+                            } else if ((dpval[i].value) == 11) {
+                                updateState(CHANNEL_TYPE_VANESUD, StringType.valueOf("SWIRL"));
+                            } else if ((dpval[i].value) == 12) {
+                                updateState(CHANNEL_TYPE_VANESUD, StringType.valueOf("WIDE"));
+                            } else {
+                                updateState(CHANNEL_TYPE_VANESUD, StringType.valueOf(String.valueOf(dpval[i].value)));
+                            }
+                            break;
+                        case 6:
+                            if ((dpval[i].value) == 0) {
+                                updateState(CHANNEL_TYPE_VANESLR, StringType.valueOf("AUTO"));
+                            } else if ((dpval[i].value) == 10) {
+                                updateState(CHANNEL_TYPE_VANESLR, StringType.valueOf("SWING"));
+                            } else if ((dpval[i].value) == 11) {
+                                updateState(CHANNEL_TYPE_VANESLR, StringType.valueOf("SWIRL"));
+                            } else if ((dpval[i].value) == 12) {
+                                updateState(CHANNEL_TYPE_VANESLR, StringType.valueOf("WIDE"));
+                            } else {
+                                updateState(CHANNEL_TYPE_VANESLR, StringType.valueOf(String.valueOf(dpval[i].value)));
+                            }
+                            break;
```
Solved with nested case

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,536 @@
+            } else {
+                logout();
+                Response resp = gson.fromJson(response, Response.class);
+                Data data = gson.fromJson(resp.data.toString(), Data.class);
```
Done

hmerk commented:
Thanks for your further comments @cpmeister . Everything is addressed.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                                        }
+                                    }
+                                    properties.put("Supported fan levels", fanLevels.toString());
+                                    updateProperties(properties);
```
Instead of calling `updateProperties` multiple times in this loop, it would be better if you only called it after you exited the loop. In fact, you might consider only calling `updateProperties` after both `populateProperties()` and this method are called as part of initialization.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+
+        // start background initialization:
+        scheduler.schedule(() -> {
+            populateProperties();
+            getDataPoints();
```
What if either of these methods fail? Would the thing properties never get populated? Perhaps you could make it retry on failure.

hmerk replied:
I don't see a reason for a retry and did not see such a thing in other bindings. If either of these fail, the thing status is set to offline, as the is some kind of communication error.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                    boolean success = resp.success;
+                    if (!success) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                        String sessionString = "{\"command\":\"login\",\"data\":{\"username\":\"Admin\",\"password\":\""
+                                + password + "\"}}";
+                        response = api.postRequest(ipAddress, sessionString);
```
`response` isn't used after this point, so why set it?

hmerk replied:
It was a leftover, removed

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    public void addChannel(String channelId, String itemType, @Nullable final Collection<?> options) {
+        if (thing.getChannel(channelId) == null) {
+            logger.trace("Channel '{}' for UID to be added", channelId);
+            ThingBuilder thingBuilder = editThing();
+            final ChannelTypeUID channelTypeUID = new ChannelTypeUID(BINDING_ID, channelId);
+            Channel channel = ChannelBuilder.create(new ChannelUID(getThing().getUID(), channelId), itemType)
+                    .withType(channelTypeUID).withKind(ChannelKind.STATE).build();
+            thingBuilder.withChannel(channel);
+            updateThing(thingBuilder.build());
+
+            if (options != null) {
+                final List<StateOption> stateOptions = options.stream()
+                        .map(e -> new StateOption(e.toString(),
+                                e.toString().substring(0, 1) + e.toString().substring(1).toLowerCase()))
+                        .collect(Collectors.toList());
```
Since you are only passing in String options you might as well make the Collection contain only Strings you don't have to call `toString()` on all of the option elements later on.
```suggestion
    public void addChannel(String channelId, String itemType, @Nullable final Collection<String> options) {
        if (thing.getChannel(channelId) == null) {
            logger.trace("Channel '{}' for UID to be added", channelId);
            ThingBuilder thingBuilder = editThing();
            final ChannelTypeUID channelTypeUID = new ChannelTypeUID(BINDING_ID, channelId);
            Channel channel = ChannelBuilder.create(new ChannelUID(getThing().getUID(), channelId), itemType)
                    .withType(channelTypeUID).withKind(ChannelKind.STATE).build();
            thingBuilder.withChannel(channel);
            updateThing(thingBuilder.build());

            if (options != null) {
                final List<StateOption> stateOptions = options.stream()
                        .map(e -> new StateOption(e,
                                e.substring(0, 1) + e.substring(1).toLowerCase()))
                        .collect(Collectors.toList());
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    public void getDataPoints() {
```
Just like `populateProperties` you should probably rename this method to `populateChannels`

hmerk replied:
Renamed

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+            }
+        }
+        if (uid != 0) {
+            login();
+            String contentString = "{\"command\":\"setdatapointvalue\",\"data\":{\"sessionID\":\"" + sessionId
+                    + "\", \"uid\":" + uid + ",\"value\":" + value + "}}";
+            String response = api.postRequest(ipAddress, contentString);
+            if (response != null) {
+                try {
+                    Response resp = gson.fromJson(response, Response.class);
+                    boolean success = resp.success;
+                    if (!success) {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                        String sessionString = "{\"command\":\"login\",\"data\":{\"username\":\"Admin\",\"password\":\""
+                                + password + "\"}}";
+                        response = api.postRequest(ipAddress, sessionString);
+                    } else {
+                        logout();
+                        updateStatus(ThingStatus.ONLINE);
+                    }
+                } catch (JsonSyntaxException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                }
+            }
+        }
```
`handleCommand` should execute and return quickly. Http requests can turn into lengthy operations so it would better if you ran this code asynchronously on the `scheduler`.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+        }
+
+        // start background initialization:
+        scheduler.schedule(() -> {
+            populateProperties();
+            getDataPoints();
+
+        }, 0, TimeUnit.SECONDS);
```
```suggestion
        scheduler.submit(() -> {
            populateProperties();
            getDataPoints();
        });
```

hmerk replied:
This gives an error in Eclipse IDE, as schedule expects three parameters:  Runnable, long, TimeUnit

cpmeister replied:
`schedule` expects three parameters, yes. But I was suggesting you changed it to `submit` instead. `execute` should also work if you prefer.

hmerk replied:
Ah, did not see that you changed it to submit. Anyway, could be changed later...

hmerk replied:
Changed

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+        }
+    }
+
+    public void login() {
```
Since `login()` can fail, you should provide a way for callers to know if it executed without issue. This can be done by either adding a return value that indicates success or by throwing an exception if something goes wrong.

Since you are currently having this method update the thing status it would be easiest to just have it return the sessionId.

```suggestion
    public String login() {
```

Then it could be called like this:
```
String sessionId = login();
if(sessionId != null){
...
    logout(sessionId);
}
```

If you want to take things further, a better way would be to refactor session management into a single method and allow callers to pass the functions that they want executed in a single session.
Lambdas make this easy.
```
private void postRequest(UnaryOperator<String> requestFactory, Consumer<Response> handler){
  String sessionId = login();
  if(sessionId != null){
    try{
      String request = requestFactory.apply(sessionId);
      String response = api.postRequest(ipAddress, contentString);
      if (response != null) {
            try {
                Response resp = gson.fromJson(response, Response.class);
                boolean success = resp.success;
                if (success) {
                   handler.accept(resp);
                } else {
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "Request unsuccessful");
                }
      }else{
         updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No Response");
      }
    } catch (JsonSyntaxException e) {
      updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
    } finally {
      logout(sessionId);
    }
  }
}
```

Then it could be called like this:
```

private void handleDataPointsResponse(Response response){
}

private void foo(){
    postRequest(sessionId -> 
       "{\"command\":\"getavailabledatapoints\",\"data\":{\"sessionID\":\"" + sessionId + "\"}}", 
        this::handleDataPointsResponse);
}
```

hmerk replied:
I have changed the sessionID handling as requested, but would like to leve your further refactoring for a later stage.
A fist try to implement this, which indeed seems to be a good idea, lead into some errors.

hmerk replied:
Implemented this as well

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                Response resp = gson.fromJson(response, Response.class);
+                boolean success = resp.success;
+                if (success) {
+                    Data data = gson.fromJson(resp.data.toString(), Data.class);
+                    Info info = gson.fromJson(data.info.toString(), Info.class);
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                Response resp = gson.fromJson(response, Response.class);
+                boolean success = resp.success;
+                if (success) {
+                    Data data = gson.fromJson(resp.data.toString(), Data.class);
+                    Id id = gson.fromJson(data.id.toString(), Id.class);
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                if (success) {
+                    Data data = gson.fromJson(resp.data.toString(), Data.class);
+                    Id id = gson.fromJson(data.id.toString(), Id.class);
+                    if (id.sessionID.toString() != null && !id.sessionID.toString().isEmpty()) {
+                        sessionId = id.sessionID.toString();
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+            try {
+                Response resp = gson.fromJson(response, Response.class);
+                boolean success = resp.success;
+                if (success) {
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                                e.toString().substring(0, 1) + e.toString().substring(1).toLowerCase()))
+                        .collect(Collectors.toList());
+                logger.trace("StateOptions : '{}'", stateOptions);
+                channel = thing.getChannel(channelId);
+                if (channel != null) {
+                    intesisStateDescriptionProvider.setStateOptions(channel.getUID(), stateOptions);
+                }
```
Removed

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                    if (success) {
+                        logout();
+                        Data data = gson.fromJson(resp.data.toString(), Data.class);
+                        Dp dp = gson.fromJson(data.dp.toString(), Dp.class);
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,529 @@
+                    }
+
+                } catch (JsonSyntaxException e) {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
Done

hmerk commented:
@cpmeister Thanks for your comments, everything addressed. Hopefully ready to merge now.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+        }
+    }
+
+    private void postRequest(UnaryOperator<String> requestFactory, Consumer<Response> handler) {
+        String sessionId = login();
+        if (sessionId != null) {
+            try {
+                String request = requestFactory.apply(sessionId);
+                logger.trace("request : '{}'", request);
+
+                String response = api.postRequest(ipAddress, request);
+                if (response != null) {
+
+                    Response resp = gson.fromJson(response, Response.class);
+                    boolean success = resp.success;
+                    if (success) {
+                        handler.accept(resp);
+                    } else {
+                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
+                                "Request unsuccessful");
+                    }
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No Response");
+                }
+            } catch (JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            } finally {
+                logout(sessionId);
+            }
+        }
+    }
```
The previous suggestion requires a version of `postRequest` that doesn't handle session state. We can simply refactor some of the inner code from old `postRequest` method to serve this purpose.
But since we don't want to confuse our two `postRequest` methods, it would also be a good idea to
rename our old `postRequest` method into something more suitable. I suggest `postRequestInSession`.
```suggestion
    
    private void postRequest(String request, Consumer<Response> handler) {
        try {
                logger.trace("request : '{}'", request);

                String response = api.postRequest(ipAddress, request);
                if (response != null) {

                    Response resp = gson.fromJson(response, Response.class);
                    boolean success = resp.success;
                    if (success) {
                        handler.accept(resp);
                    } else {
                        updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR,
                                "Request unsuccessful");
                    }
                } else {
                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, "No Response");
                }
         } catch (JsonSyntaxException e) {
                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
         }
    }

    private void postRequestInSession(UnaryOperator<String> requestFactory, Consumer<Response> handler) {
        String sessionId = login();
        if (sessionId != null) {
            try {
                String request = requestFactory.apply(sessionId);
                postRequest(request, handler);
            } finally {
                logout(sessionId);
            }
        }
    }
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    public void populateProperties() {
+        String contentString = "{\"command\":\"getinfo\",\"data\":\"\"}";
+        String response = api.postRequest(ipAddress, contentString);
+        logger.trace("getInfo response : {}", response);
+        if (response != null) {
+            try {
+                Response resp = gson.fromJson(response, Response.class);
+                boolean success = resp.success;
+                if (success) {
+                    Data data = gson.fromJson(resp.data, Data.class);
+                    Info info = gson.fromJson(data.info, Info.class);
+                    properties.put(PROPERTY_VENDOR, "Intesis");
+                    properties.put(PROPERTY_MODEL_ID, info.deviceModel);
+                    properties.put(PROPERTY_SERIAL_NUMBER, info.sn);
+                    properties.put(PROPERTY_FIRMWARE_VERSION, info.fwVersion);
+                    properties.put(PROPERTY_MAC_ADDRESS, info.wlanSTAMAC);
+                    updateStatus(ThingStatus.ONLINE);
+                }
+            } catch (JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            }
+        }
```
This code can now also be refactored to use our new `postRequest` method.
```suggestion
        postRequest("{\"command\":\"getinfo\",\"data\":\"\"}", resp -> {
                    Data data = gson.fromJson(resp.data, Data.class);
                    Info info = gson.fromJson(data.info, Info.class);
                    properties.put(PROPERTY_VENDOR, "Intesis");
                    properties.put(PROPERTY_MODEL_ID, info.deviceModel);
                    properties.put(PROPERTY_SERIAL_NUMBER, info.sn);
                    properties.put(PROPERTY_FIRMWARE_VERSION, info.fwVersion);
                    properties.put(PROPERTY_MAC_ADDRESS, info.wlanSTAMAC);
                    updateStatus(ThingStatus.ONLINE);
        });
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+
+            if (options != null) {
+                final List<StateOption> stateOptions = options.stream()
+                        .map(e -> new StateOption(e, e.substring(0, 1) + e.toString().substring(1).toLowerCase()))
```
```suggestion
                        .map(e -> new StateOption(e, e.substring(0, 1) + e.substring(1).toLowerCase()))
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+        ipAddress = config.ipAddress;
+        password = config.password;
+        if (ipAddress.isEmpty() || password.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR);
+        }
```
If configuration is invalid there is no reason to continue initializing right?
Also, you should specify in the status details which configuration parameter is wrong so that users would
know how to fix it.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        final IntesisConfiguration config = getConfigAs(IntesisConfiguration.class);
+        ipAddress = config.ipAddress;
+        password = config.password;
```
There is no need to cache these as their own fields in the handler when you could always just retrieve them directly from the config each time.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+    @Override
+    public void initialize() {
+        updateStatus(ThingStatus.UNKNOWN);
+        final IntesisConfiguration config = getConfigAs(IntesisConfiguration.class);
```
I assume it was your original intention save the config in your `config` field. This is the standard design pattern in most bindings.
```suggestion
        config = getConfigAs(IntesisConfiguration.class);
```

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+
+    private final Map<String, String> properties = new HashMap<>();
+
+    private IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider;
```
Might as well make this final.
```suggestion
    private final IntesisDynamicStateDescriptionProvider intesisStateDescriptionProvider;
```
Also, please move all final fields above the non-final ones.

hmerk replied:
Done

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+    private String ipAddress = "";
+    private String password = "";
+
+    final Gson gson = new Gson();
```
```suggestion
    private final Gson gson = new Gson();
```

hmerk replied:
Done

cpmeister reviewed (changes requested):
These are my last change requests and then I'll accept and then merge this.

Thanks for all the work you have done on this!

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+            final int uId = uid;
+            final int newValue = value;
+            scheduler.submit(() -> {
+                login();
+                String sessionId = login();
+                if (sessionId != null) {
+                    String contentString = "{\"command\":\"setdatapointvalue\",\"data\":{\"sessionID\":\"" + sessionId
+                            + "\", \"uid\":" + uId + ",\"value\":" + newValue + "}}";
+                    String response = api.postRequest(ipAddress, contentString);
+                    if (response != null) {
+                        try {
+                            Response resp = gson.fromJson(response, Response.class);
+                            boolean success = resp.success;
+                            if (!success) {
+                                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                            } else {
+                                logout(sessionId);
+                                updateStatus(ThingStatus.ONLINE);
+                            }
+                        } catch (JsonSyntaxException e) {
+                            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                        }
+                    }
+                }
```
Done

hmerk commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -0,0 +1,537 @@
+    }
+
+    public @Nullable String login() {
+        String contentString = "{\"command\":\"login\",\"data\":{\"username\":\"Admin\",\"password\":\"" + password
+                + "\"}}";
+        String response = api.postRequest(ipAddress, contentString);
+        String sessionId = null;
+        if (response != null) {
+            try {
+                Response resp = gson.fromJson(response, Response.class);
+                boolean success = resp.success;
+                if (success) {
+                    Data data = gson.fromJson(resp.data, Data.class);
+                    Id id = gson.fromJson(data.id, Id.class);
+                    sessionId = id.sessionID.toString();
+                    if (sessionId != null && !sessionId.isEmpty()) {
+                        updateStatus(ThingStatus.ONLINE);
+                        return sessionId;
+                    } else {
+                        updateStatus(ThingStatus.OFFLINE);
+                    }
+                } else {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+                    return null;
+                }
+            } catch (JsonSyntaxException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getMessage());
+            }
+        } else {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
+            return null;
+        }
+        return sessionId;
```
Done

hmerk commented:
Thanks @cpmeister, I am done with your last change requests.

cpmeister commented on bundles/org.openhab.binding.intesis/src/main/java/org/openhab/binding/intesis/internal/handler/IntesisHomeHandler.java:
```diff
@@ -99,21 +97,22 @@ public IntesisHomeHandler(final Thing thing, final HttpClient httpClient,
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "IP-Address not set");
+        } else if (config.password.isEmpty()) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Password not set");
         }
```
Once you find that the configuration is invalid, you should bail out early subsequent errors from the improper configuration would end up burying the initial config errors.
```suggestion
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "IP-Address and password not set");
            return;
        } else if (config.ipAddress.isEmpty()) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "IP-Address not set");
            return;
        } else if (config.password.isEmpty()) {
            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, "Password not set");
            return;
        }
```

Hilbrand replied:
Maybe better to add an else case at the last if and put the rest of the code in there
```
 } else {
    // start background initialization:
   ....
}
```

hmerk replied:
Done, implemented both changes.

Hilbrand replied:
With the else the return is technically not needed :wink: :

cpmeister reviewed (changes requested):
I mentioned this in the previous requests but looks like it was missed.
Final change!

hmerk commented:
Thanks @cpmeister and @Hilbrand , done with the last changes.

cpmeister commented:
@fwolter Can you take another look and check if your requests have been satisfied?

fwolter commented:
Feel free to merge.

hmerk commented:
Thanks @Hilbrand @cpmeister @fwolter for your great support 👍
